lib/concurrent/atom.rb

```diff
@@ -0,0 +1,136 @@
+require 'concurrent/dereferenceable'
+require 'concurrent/atomic/atomic_reference'
+require 'concurrent/synchronization/object'
+
+module Concurrent
+
+  # Atoms provide a way to manage shared, synchronous, independent state.
+  #
+  # An atom is initialized with an initial value and an optional validation
+  # proc. At any time the value of the atom can be synchronously and safely
+  # changed. If a validator is given at construction then any new value
+  # will be checked against the validator and will be rejected if the
+  # validator returns false or raises an exception.
+  #
+  # There are two ways to change the value of an atom: {#compare_and_set} and
+  # {#swap}. The former will set the new value if and only if it validates and
+  # the current value matches the new value. The latter will atomically set the
+  # new value to the result of running the given block if and only if that
+  # value validates.
+  #
+  # @!macro copy_options
+  #
+  # @see http://clojure.org/atoms Clojure Atoms
+  class Atom < Synchronization::Object
+    include Dereferenceable
+
+    # @!macro [attach] atom_initialize
+    #
+    #   Create a new atom with the given initial value.
+    #
+    #   @param [Object] value The initial value
+    #   @param [Hash] opts The options used to configure the atom
+    #   @option opts [Proc] :validator (nil) Optional proc used to validate new
+    #     values. It must accept one and only one argument which will be the
+    #     intended new value. The validator will return true if the new value
+    #     is acceptable else return false (preferrably) or raise an exception.
+    #
+    #   @!macro deref_options
+    # 
+    #   @raise [ArgumentError] if the validator is not a `Proc` (when given)
+    def initialize(value, opts = {})
+      super()
+      synchronize{ ns_initialize(value, opts) }
```

Comment:
it can be either synchronized but then all later access has to be synchronized too, which is not what you want probably. `ensure_ivar_visibility!` will ensure visibility but later access synchronization is not required.